### PR TITLE
Add "placeholder" (pre-MVP) implementation of osu!catch editor

### PR DESCRIPTION
**Pull request by ekrctb**
It is a set of files that are needed for any plan of the catch editor anyway.
Due to complexity, juice stream editing/placement is not implemented in this PR.

### Known bugs (I expect this PR to be merged with known issues)

- Catcher size is not changed when the circle size is changed
- Scrolling speed is not changed when the approach rate is changed
- Objects can go out of bounds (only tiny droplets are confined to the playfield)
- Catcher trails/hit lighting are duplicated. Catcher animation is not rewound correctly.
- Fruit outline is always a circle even with a non-default skin.

Fixes to some issues are not included in this PR because that requires changes to the `osu.Game` code or non-editor parts.
Others are not implemented yet due to complexity.

### Not implemented (features needed even for basic editing)

- Editing/placement of juice streams
  - At least piecewise-linear ones.
- Position snapping
- Copy/paste (serialization)
- Probably a lot more
- Hyper state of hit objects should be updated real-time when position is changed <https://github.com/ppy/osu/pull/13617#issuecomment-865827951>
- Hit object shouldn't be expired when caught by the catcher <https://github.com/ppy/osu/pull/13617#issuecomment-866247870>

https://user-images.githubusercontent.com/191335/124296661-2e665300-db95-11eb-8da5-ad34241a6e19.mp4

### CODE REVIEW

File: osu.Game.Rulesets.Catch/Edit/Blueprints/Components/FruitOutline.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+namespace osu.Game.Rulesets.Catch.Edit.Blueprints.Components
+{
+    public class FruitOutline : CompositeDrawable
+    {
+        public FruitOutline()
+        {
+            Anchor = Anchor.BottomLeft;
+            Origin = Anchor.Centre;
+            Size = new Vector2(2 * CatchHitObject.OBJECT_RADIUS);
+            // TODO: use skinned component?
```
Comment (peppy) [inline osu.Game.Rulesets.Catch/Edit/Blueprints/Components/FruitOutline.cs:22]: Just for reference, in general we're aiming to keep the editor blueprints as design/skin agnostic. They should exist as a layer that is placed over the top of the actual gameplay, and can be toggled at any point in time. So what you have is probably more the direction we'd be going, like in other rulesets' editor implementations.

**Comment by peppy** [comment]
One other issue to add to the list is that hyperdash state doesn't update on `X` movement of fruits (only `Y` / `Time` changes trigger an update currently).

**Review by smoogipoo — DISMISSED** [review]
Looks fine to me as an initial pass :)

File: osu.Game.Rulesets.Catch/Edit/BananaShowerCompositionTool.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Game.Rulesets.Catch.Edit.Blueprints;
+using osu.Game.Rulesets.Catch.Objects;
+using osu.Game.Rulesets.Edit;
+using osu.Game.Rulesets.Edit.Tools;
+
+namespace osu.Game.Rulesets.Catch.Edit
+{
+    public class BananaShowerCompositionTool : HitObjectCompositionTool
```
Comment (bdach) [inline osu.Game.Rulesets.Catch/Edit/BananaShowerCompositionTool.cs:11]: May also want to override `CreateIcon()` - other rulesets do

```csharp
public override Drawable CreateIcon() => new BeatmapStatisticIcon(BeatmapStatisticsIconType.Spinners);
```

File: osu.Game.Rulesets.Catch/Edit/FruitCompositionTool.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Game.Rulesets.Catch.Edit.Blueprints;
+using osu.Game.Rulesets.Catch.Objects;
+using osu.Game.Rulesets.Edit;
+using osu.Game.Rulesets.Edit.Tools;
+
+namespace osu.Game.Rulesets.Catch.Edit
+{
+    public class FruitCompositionTool : HitObjectCompositionTool
```
Comment (bdach) [inline osu.Game.Rulesets.Catch/Edit/FruitCompositionTool.cs:11]: similarly as above here

```csharp
public override Drawable CreateIcon() => new BeatmapStatisticIcon(BeatmapStatisticsIconType.Circles);
```

File: osu.Game/Screens/Edit/Compose/Components/ComposeBlueprintContainer.cs [hunk 4]
```diff
@@ -256,7 +256,7 @@ protected sealed override SelectionBlueprint<HitObject> CreateBlueprintFor(HitOb
             if (drawable == null)
                 return null;
 
-            return CreateHitObjectBlueprintFor(item).With(b => b.DrawableObject = drawable);
+            return CreateHitObjectBlueprintFor(item)?.With(b => b.DrawableObject = drawable);
         }
 
         public virtual HitObjectSelectionBlueprint CreateHitObjectBlueprintFor(HitObject hitObject) => null;
```
Comment (bdach) [inline osu.Game/Screens/Edit/Compose/Components/ComposeBlueprintContainer.cs:262]: may as well mark as `[CanBeNull]`, if we're doing the change above

**Review by bdach — COMMENTED** [review]
A few minor things that may as well be fixed here.

**Comment by bdach** [comment]
Another issue for the pile: Timeline blueprints pop out of existence shortly after their corresponding objects are caught in the playfield, unless they are currently selected. (At least they pop back up on rewind, though.)

**Review by peppy — APPROVED** [review]

**Comment by peppy** [comment]
This looks great as a first implementation.

One more thing which should be considered a relatively high priority TODO is adding a time snap grid (basically bar lines on the playfield). Could be implemented similarly to mania's, I think.

20210623 135840 (dotnet)

**Pull request merged**